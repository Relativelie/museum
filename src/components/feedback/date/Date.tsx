import { FC, useState, useEffect, FormEvent } from 'react';
import {
    currentDate,
    currentMonth,
    currentYear,
    futureDate,
    futureMonth,
    futureYear }
    from './dateValues';

export const Date: FC = () => {
    const [selectedDate, setSelectedDate] = useState('');
    useEffect(() => {
        const date = `${currentYear}-${checkMonthAndDay(currentMonth)}-${checkMonthAndDay(currentDate)}`;
        setSelectedDate(date);
    }, []);

    const checkMonthAndDay = (value: number) => {
        return value.toString().length === 2 ? value : `0${value}`;
    };

    const selectDate = (e: FormEvent<HTMLInputElement>) => {
        const elem = e.target as HTMLInputElement;
        const selectedValue = elem.valueAsDate;
        if (selectedValue !== null) {
            const myMonth = selectedValue.getMonth() + 1;
            const selectedMonth = checkMonthAndDay(myMonth);
            const selectedDay = checkMonthAndDay(selectedValue.getDate());
            const result = `${selectedValue.getFullYear()}-${selectedMonth}-${selectedDay}`;
            setSelectedDate(result);
        }
    };

    return (
        <input
            className="feedback_input"
            name="date"
            type="date"
            data-testid="form-field-date"
            value={selectedDate}
            onChange={(e) => selectDate(e)}
            min={`${currentYear}-${checkMonthAndDay(currentMonth)}-${currentDate}`}
            max={`${futureYear}-${checkMonthAndDay(futureMonth)}-${futureDate}`}
        />
    );
};
