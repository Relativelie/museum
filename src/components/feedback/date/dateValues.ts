const current = new Date();
export const currentDate = current.getDate();
export const currentMonth = current.getMonth() + 1;
export const currentYear = current.getFullYear();

const numberOfDaysToAdd = 10;
const futureNumber = current.setDate(current.getDate() + numberOfDaysToAdd);
const future = new Date(futureNumber);
export const futureDate = future.getDate();
export const futureMonth = future.getMonth() + 1;
export const futureYear = future.getFullYear();
