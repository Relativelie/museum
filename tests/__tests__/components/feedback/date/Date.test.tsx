import '@testing-library/jest-dom/extend-expect';
import '@testing-library/jest-dom';
import { act, create } from 'react-test-renderer';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { screen } from '@testing-library/dom';
import { Date } from '../../../../../src/components/feedback/date/Date';

describe('Feedback - Date component', () => {
    test('snapshot - date component', () => {
        let root: any;
        act(() => {
            root = create(<Date />);
        });
        expect(root.toJSON()).toMatchSnapshot();
    });

    test('change date value', () => {
        render(<Date />);
        const inputEl = screen.getByTestId('form-field-date');
        userEvent.type(inputEl, '2022-04-01');
        expect(screen.getByTestId('form-field-date')).toHaveValue('2022-04-01');
    });
});
