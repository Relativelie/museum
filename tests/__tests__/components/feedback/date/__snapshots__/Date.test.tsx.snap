// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Feedback - Date component snapshot - date component 1`] = `
<input
  className="feedback_input"
  max="2022-04-28"
  min="2022-04-18"
  name="date"
  onChange={[Function]}
  type="date"
  value="2022-04-18"
/>
`;
